src/lib/components/TableOfContents/index.js

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {html} from 'lit-element';
+import {BaseStateElement} from '../BaseStateElement';
+import {closeToC} from '../../actions';
+
+/**
+ * Element that renders table of contents.
+ * @extends {BaseStateElement}
+ * @final
+ */
+class TableOfContents extends BaseStateElement {
+  static get properties() {
+    return {
+      title: {type: String},
+      titleId: {type: String, attribute: 'title-id'},
+      opened: {type: Boolean, reflect: true},
+    };
+  }
+
+  constructor() {
+    super();
+    this.scrollSpy = this.scrollSpy.bind(this);
+    this.openedToFalse = this.openedToFalse.bind(this);
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.tocHTML = this.innerHTML;
+    this.divContent =
+      document.querySelector('#content') || document.createElement('div');
+    this.headings = this.divContent.querySelectorAll('h1[id], h2[id], h3[id]');
+
+    this.observer = new IntersectionObserver(this.scrollSpy);
+    this.headings.forEach((heading) => this.observer.observe(heading));
```

Comment:
I'd suggest looking into if you can use the rootMargin option of the observer to create a smaller "window" only at the top of the page that you're spying. Otherwise you're spying every header on the page and that means as soon as the article opens it's already the heading at the bottom.

![image](https://user-images.githubusercontent.com/1066253/87481291-e5b03580-c5e3-11ea-9835-723dcc36df70.png)